### ISSUE TITLE
[Android] Unable to read from database HITS - crashes on startup

### ISSUE THREAD

**Issue by anthonymonori** [Adobe-Marketing-Cloud/mobile-services#200]
I was getting a crash on startup of my application while testing device space edge-cases. When I set my device to full (e.g. there would be no more free-space to use), I've got the following log on 1st startup:
```
E: (1) no such table: HITS
E: ADBMobile Error : Analytics - Unable to read from database (no such table: HITS (code 1): , while compiling: SELECT ID, URL, TIMESTAMP FROM HITS ORDER BY ID ASC LIMIT 1)
E: FATAL EXCEPTION: AsyncTask #2
   Process: _redacted_, PID: 2566
   java.lang.RuntimeException: An error occured while executing doInBackground()
       at android.os.AsyncTask$3.done(AsyncTask.java:304)
       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:355)
       at java.util.concurrent.FutureTask.setException(FutureTask.java:222)
       at java.util.concurrent.FutureTask.run(FutureTask.java:242)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818)
    Caused by: android.database.sqlite.SQLiteException: Failed to change locale for db '/data/user/0/_redacted_/databases/CommentsDb' to 'en_US'.
       at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:181)
       at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:137)
       at android.content.ContentProviderProxy.query(ContentProviderNative.java:421)
       at android.content.ContentResolver.query(ContentResolver.java:478)
       at android.content.CursorLoader.loadInBackground(CursorLoader.java:64)
       at android.content.CursorLoader.loadInBackground(CursorLoader.java:42)
       at android.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:312)
       at android.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:69)
       at android.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:57)
       at android.os.AsyncTask$2.call(AsyncTask.java:292)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
       at java.lang.Thread.run(Thread.java:818) 
E: Couldn't create directory for SharedPreferences file /data/data/_dedacted_/shared_prefs/APP_MEASUREMENT_CACHE.xml
```

After that, I gave it a second go, but still crashed:
```
W: writeToFile: Got exception:
   java.io.IOException: write failed: ENOSPC (No space left on device)
       at libcore.io.IoBridge.write(IoBridge.java:502)
       at java.io.FileOutputStream.write(FileOutputStream.java:186)
       at com.android.internal.util.FastXmlSerializer.flushBytes(FastXmlSerializer.java:232)
       at com.android.internal.util.FastXmlSerializer.flush(FastXmlSerializer.java:253)
       at com.android.internal.util.FastXmlSerializer.endDocument(FastXmlSerializer.java:198)
       at com.android.internal.util.XmlUtils.writeMapXml(XmlUtils.java:188)
       at android.app.SharedPreferencesImpl.writeToFile(SharedPreferencesImpl.java:597)
       at android.app.SharedPreferencesImpl.access$800(SharedPreferencesImpl.java:51)
       at android.app.SharedPreferencesImpl$2.run(SharedPreferencesImpl.java:512)
       at android.app.SharedPreferencesImpl.enqueueDiskWrite(SharedPreferencesImpl.java:533)
       at android.app.SharedPreferencesImpl.access$100(SharedPreferencesImpl.java:51)
       at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java:455)
       at com.adobe.mobile.StaticMethods.getAID(StaticMethods.java:697)
       at com.adobe.mobile.RequestBuilder.buildAndSendRequest(RequestBuilder.java:56)
       at com.adobe.mobile.AnalyticsTrackInternal.trackInternal(AnalyticsTrackInternal.java:41)
       at com.adobe.mobile.Lifecycle.start(Lifecycle.java:173)
       at com.adobe.mobile.Config$10.run(Config.java:237)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818)
    Caused by: android.system.ErrnoException: write failed: ENOSPC (No space left on device)
       at libcore.io.Posix.writeBytes(Native Method)
       at libcore.io.Posix.write(Posix.java:258)
       at libcore.io.BlockGuardOs.write(BlockGuardOs.java:313)
       at libcore.io.IoBridge.write(IoBridge.java:497)
       at java.io.FileOutputStream.write(FileOutputStream.java:186) 
       at com.android.internal.util.FastXmlSerializer.flushBytes(FastXmlSerializer.java:232) 
       at com.android.internal.util.FastXmlSerializer.flush(FastXmlSerializer.java:253) 
       at com.android.internal.util.FastXmlSerializer.endDocument(FastXmlSerializer.java:198) 
       at com.android.internal.util.XmlUtils.writeMapXml(XmlUtils.java:188) 
       at android.app.SharedPreferencesImpl.writeToFile(SharedPreferencesImpl.java:597) 
       at android.app.SharedPreferencesImpl.access$800(SharedPreferencesImpl.java:51) 
       at android.app.SharedPreferencesImpl$2.run(SharedPreferencesImpl.java:512) 
       at android.app.SharedPreferencesImpl.enqueueDiskWrite(SharedPreferencesImpl.java:533) 
       at android.app.SharedPreferencesImpl.access$100(SharedPreferencesImpl.java:51) 
       at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java:455) 
       at com.adobe.mobile.StaticMethods.getAID(StaticMethods.java:697) 
       at com.adobe.mobile.RequestBuilder.buildAndSendRequest(RequestBuilder.java:56) 
       at com.adobe.mobile.AnalyticsTrackInternal.trackInternal(AnalyticsTrackInternal.java:41) 
       at com.adobe.mobile.Lifecycle.start(Lifecycle.java:173) 
       at com.adobe.mobile.Config$10.run(Config.java:237) 
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
       at java.lang.Thread.run(Thread.java:818) 
E: (13) statement aborts at 62: [CREATE TABLE IF NOT EXISTS HITS (ID INTEGER PRIMARY KEY AUTOINCREMENT, URL TEXT, TIMESTAMP INTEGER)] database or disk is full
E: ADBMobile Error : Analytics - Unable to create database due to a sql error (database or disk is full (code 13))
E: (1) no such table: HITS
E: ADBMobile Error : Analytics - Unable to create database due to a sql error (no such table: HITS (code 1): , while compiling: INSERT INTO HITS (URL, TIMESTAMP) VALUES (?, ?))
E: (1) no such table: HITS
E: ADBMobile Error : Analytics - Unable to get tracking queue size due to a sql error (no such table: HITS (code 1): , while compiling: select count(*) from HITS)
E: ADBMobile Error : Analytics - Unknown error while inserting url (ndh=1&ce=UTF-8&pev2=ADBINTERNAL%3ALifecycle&c.&a.&RunMode=Application&HourOfDay=22&Resolution=1536x1952&MonthlyEngUserEvent=MonthlyEngUserEvent&InstallEvent=InstallEvent&DailyEngUserEvent=DailyEngUserEvent&InstallDate=1%2F23%2F2017&CarrierName=Android&AppID=LEGO%20Life%201.1.0%20%2880000012%29&OSVersion=Android%205.1.1&Launches=1&DeviceName=Android%20SDK%20built%20for%20x86&DayOfWeek=2&LaunchEvent=LaunchEvent&internalaction=Lifecycle&.a&.c&t=00%2F00%2F0000%2000%3A00%3A00%200%20-60&pageName=LEGO%20Life%201.1.0%20%2880000012%29&aid=2C4339A80530B75B-40000300800171DB&pe=lnk_o)
E: ADBMobile Error : Analytics - Database in unrecoverable state (Attempt to invoke virtual method 'void android.database.sqlite.SQLiteStatement.bindString(int, java.lang.String)' on a null object reference), resetting.
E: (28) file unlinked while open: /data/data/_redacted_/cache/ADBMobileDataCache.sqlite
E: (13) statement aborts at 62: [CREATE TABLE IF NOT EXISTS HITS (ID INTEGER PRIMARY KEY AUTOINCREMENT, URL TEXT, TIMESTAMP INTEGER)] database or disk is full
E: ADBMobile Error : Analytics - Unable to create database due to a sql error (database or disk is full (code 13))
E: (1) no such table: HITS
E: ADBMobile Error : Analytics - Unable to create database due to a sql error (no such table: HITS (code 1): , while compiling: INSERT INTO HITS (URL, TIMESTAMP) VALUES (?, ?))
W: writeToFile: Got exception:
   java.io.IOException: write failed: ENOSPC (No space left on device)
       at libcore.io.IoBridge.write(IoBridge.java:502)
       at java.io.FileOutputStream.write(FileOutputStream.java:186)
       at com.android.internal.util.FastXmlSerializer.flushBytes(FastXmlSerializer.java:232)
       at com.android.internal.util.FastXmlSerializer.flush(FastXmlSerializer.java:253)
       at com.android.internal.util.FastXmlSerializer.endDocument(FastXmlSerializer.java:198)
       at com.android.internal.util.XmlUtils.writeMapXml(XmlUtils.java:188)
       at android.app.SharedPreferencesImpl.writeToFile(SharedPreferencesImpl.java:597)
       at android.app.SharedPreferencesImpl.access$800(SharedPreferencesImpl.java:51)
       at android.app.SharedPreferencesImpl$2.run(SharedPreferencesImpl.java:512)
       at android.app.SharedPreferencesImpl.enqueueDiskWrite(SharedPreferencesImpl.java:533)
       at android.app.SharedPreferencesImpl.access$100(SharedPreferencesImpl.java:51)
       at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java:455)
       at com.adobe.mobile.Lifecycle.resetLifecycleFlags(Lifecycle.java:382)
       at com.adobe.mobile.Lifecycle.start(Lifecycle.java:185)
       at com.adobe.mobile.Config$10.run(Config.java:237)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818)
    Caused by: android.system.ErrnoException: write failed: ENOSPC (No space left on device)
       at libcore.io.Posix.writeBytes(Native Method)
       at libcore.io.Posix.write(Posix.java:258)
       at libcore.io.BlockGuardOs.write(BlockGuardOs.java:313)
       at libcore.io.IoBridge.write(IoBridge.java:497)
       at java.io.FileOutputStream.write(FileOutputStream.java:186) 
       at com.android.internal.util.FastXmlSerializer.flushBytes(FastXmlSerializer.java:232) 
       at com.android.internal.util.FastXmlSerializer.flush(FastXmlSerializer.java:253) 
       at com.android.internal.util.FastXmlSerializer.endDocument(FastXmlSerializer.java:198) 
       at com.android.internal.util.XmlUtils.writeMapXml(XmlUtils.java:188) 
       at android.app.SharedPreferencesImpl.writeToFile(SharedPreferencesImpl.java:597) 
       at android.app.SharedPreferencesImpl.access$800(SharedPreferencesImpl.java:51) 
       at android.app.SharedPreferencesImpl$2.run(SharedPreferencesImpl.java:512) 
       at android.app.SharedPreferencesImpl.enqueueDiskWrite(SharedPreferencesImpl.java:533) 
       at android.app.SharedPreferencesImpl.access$100(SharedPreferencesImpl.java:51) 
       at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java:455) 
       at com.adobe.mobile.Lifecycle.resetLifecycleFlags(Lifecycle.java:382) 
       at com.adobe.mobile.Lifecycle.start(Lifecycle.java:185) 
       at com.adobe.mobile.Config$10.run(Config.java:237) 
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
       at java.lang.Thread.run(Thread.java:818) 
E: ADBMobile Error : Analytics - Unknown error while inserting url (ndh=1&ce=UTF-8&c.&a.&DeviceName=Android%20SDK%20built%20for%20x86&OSVersion=Android%205.1.1&RunMode=Application&Resolution=1536x1952&CarrierName=Android&AppID=LEGO%20Life%201.1.0%20%2880000012%29&.a&LoggedIn=0&appname=lego%20life&appversion=1&AnalyticsAnonSessionId=1025282e-5570-485b-84d0-3a81cbfa86ea&.c&t=00%2F00%2F0000%2000%3A00%3A00%200%20-60&pageName=start_screen&aid=2C4339A80530B75B-40000300800171DB)
E: ADBMobile Error : Analytics - Database in unrecoverable state (Attempt to invoke virtual method 'void android.database.sqlite.SQLiteStatement.bindString(int, java.lang.String)' on a null object reference), resetting.
E: (28) file unlinked while open: /data/data/_redacted_/cache/ADBMobileDataCache.sqlite
E: (13) statement aborts at 62: [CREATE TABLE IF NOT EXISTS HITS (ID INTEGER PRIMARY KEY AUTOINCREMENT, URL TEXT, TIMESTAMP INTEGER)] database or disk is full
E: ADBMobile Error : Analytics - Unable to create database due to a sql error (database or disk is full (code 13))
E: (1) no such table: HITS
E: ADBMobile Error : Analytics - Unable to create database due to a sql error (no such table: HITS (code 1): , while compiling: INSERT INTO HITS (URL, TIMESTAMP) VALUES (?, ?))
W: writeToFile: Got exception:
   java.io.IOException: write failed: ENOSPC (No space left on device)
       at libcore.io.IoBridge.write(IoBridge.java:502)
       at java.io.FileOutputStream.write(FileOutputStream.java:186)
       at com.android.internal.util.FastXmlSerializer.flushBytes(FastXmlSerializer.java:232)
       at com.android.internal.util.FastXmlSerializer.flush(FastXmlSerializer.java:253)
       at com.android.internal.util.FastXmlSerializer.endDocument(FastXmlSerializer.java:198)
       at com.android.internal.util.XmlUtils.writeMapXml(XmlUtils.java:188)
       at android.app.SharedPreferencesImpl.writeToFile(SharedPreferencesImpl.java:597)
       at android.app.SharedPreferencesImpl.access$800(SharedPreferencesImpl.java:51)
       at android.app.SharedPreferencesImpl$2.run(SharedPreferencesImpl.java:512)
       at android.app.SharedPreferencesImpl.enqueueDiskWrite(SharedPreferencesImpl.java:533)
       at android.app.SharedPreferencesImpl.access$100(SharedPreferencesImpl.java:51)
       at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java:455)
       at com.adobe.mobile.Lifecycle.stop(Lifecycle.java:198)
       at com.adobe.mobile.Config$12.run(Config.java:264)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818)
    Caused by: android.system.ErrnoException: write failed: ENOSPC (No space left on device)
       at libcore.io.Posix.writeBytes(Native Method)
       at libcore.io.Posix.write(Posix.java:258)
       at libcore.io.BlockGuardOs.write(BlockGuardOs.java:313)
       at libcore.io.IoBridge.write(IoBridge.java:497)
       at java.io.FileOutputStream.write(FileOutputStream.java:186) 
       at com.android.internal.util.FastXmlSerializer.flushBytes(FastXmlSerializer.java:232) 
       at com.android.internal.util.FastXmlSerializer.flush(FastXmlSerializer.java:253) 
       at com.android.internal.util.FastXmlSerializer.endDocument(FastXmlSerializer.java:198) 
       at com.android.internal.util.XmlUtils.writeMapXml(XmlUtils.java:188) 
       at android.app.SharedPreferencesImpl.writeToFile(SharedPreferencesImpl.java:597) 
       at android.app.SharedPreferencesImpl.access$800(SharedPreferencesImpl.java:51) 
       at android.app.SharedPreferencesImpl$2.run(SharedPreferencesImpl.java:512) 
       at android.app.SharedPreferencesImpl.enqueueDiskWrite(SharedPreferencesImpl.java:533) 
       at android.app.SharedPreferencesImpl.access$100(SharedPreferencesImpl.java:51) 
       at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java:455) 
       at com.adobe.mobile.Lifecycle.stop(Lifecycle.java:198) 
       at com.adobe.mobile.Config$12.run(Config.java:264) 
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
       at java.lang.Thread.run(Thread.java:818) 
```

I am currently using **Android SDK v4.13.2** and I can see in **v4.13.3** you have noted in the changelog "Analytics - Improved handling of read-only database access". Before I take the long route to upgrade the Adobe SDK to it, could you confirm that it is regarding the same issue I am experiencing here?

**Bonus question:** Also, any chance of going open source with these mobile sdks at some point? :)

**Comment by shalehaha** [comment]
@anthonymonori,

The crash was caused by accessing SharedPreferences when the disk is full, so it is a different issue than what we have fixed in v4.13.3. I've filed a bug in our internal bug tracking system and will update this thread once we have a release date for the fix. Thanks. 

And I don't think it is possible to open source the SDK in the near future :)

Jiabin

**Comment by anthonymonori** [comment]
@shalehaha Is this something you investigated further? I will update to v4.13.4 and report back as well.

**Comment by shalehaha** [comment]
@anthonymonori,
It is scheduled to be fixed in the April release which comes out around April 20th.

**Comment by shalehaha** [comment]
@anthonymonori,

Sorry for the late response, but when I looked into this issue, I found that it won't cause the crash of the SDK. When you look at the Android code, https://android.googlesource.com/platform/frameworks/base/+/android-3.2.4_r1/core/java/android/app/SharedPreferencesImpl.java#591, they have already try-catch the IOException, and print the log for it. So what you have pasted is just the debug log, none of them is actually the reason of the crash. 
I also manually tried filling up the disk and ran the app, and it was able to run without crash.

So I guess there is something else makes your app crashing.
